src/eddsa.rs

```diff
@@ -78,6 +82,99 @@ pub struct PrivateKey<E: JubjubEngine>(pub E::Fs);
 #[derive(Clone)]
 pub struct PublicKey<E: JubjubEngine>(pub Point<E, Unknown>);
 
+#[derive(Clone)]
+pub struct Seed<E: JubjubEngine>(pub E::Fs);
+
+impl<E: JubjubEngine> Seed<E> {
+    pub fn random_seed<R: Rng>(rng: &mut R, msg: &[u8]) -> Self {
+        // T = (l_H + 128) bits of randomness
+        // For H*, l_H = 512 bits
+        let mut t = [0u8; 80];
+        rng.fill_bytes(&mut t[..]);
+
+        // Generate randomness using hash function based on some entropy and the message
+        // Generation of randommess is completely off-chain, so we use BLAKE2b!
+        // r = H*(T || M)
+        Seed(h_star::<E>(&t[..], msg))
+    }
+
+    pub fn deterministic_seed(pk: &PrivateKey<E>, msg: &[u8]) -> Self {
+        use std::slice;
+
+        let mut hasher = Sha256::new();
+        hasher.input(msg);
+        let h1 = hasher.result(); // h1 = sha256(msg)
+
+        let zero = [0x00];
+        let one = [0x01];
+        let mut v = [1u8; 32];  // v = 0x01 0x01 0x01 ... 0x01
+        let mut key = [0u8; 32]; //  key = 0x00 0x00 0x00 ... 0x00
+
+        let mut mac = HmacSha256::new_varkey(&key).expect("HMAC can take key of any size");
+        
+        // concatenated = v || 0x00 || priv_key || h1
+        let mut concatenated: Vec<u8> = v.as_ref().to_vec();
+        concatenated.extend(zero.as_ref().to_vec().into_iter());
+        let priv_key: Vec<[u8; 8]> = pk.0
+            .into_repr()
+            .as_ref()
+            .to_vec()
+            .into_iter()
+            .map(|x| x.to_be_bytes())
+            .collect();
+        concatenated.extend(priv_key.as_slice().join(&0).into_iter());
+        concatenated.extend(h1.as_slice().to_vec().into_iter());
+
+        mac.input(&concatenated[..]);
+        key.copy_from_slice(mac.clone().result().code().as_mut_slice()); // key = HMAC(key, concatenated)
+
+        mac = HmacSha256::new_varkey(&key).expect("HMAC can take key of any size");
+        mac.input(&v);
+        v.copy_from_slice(mac.clone().result().code().as_mut_slice()); // v = HMAC(key, v)
+
+        // concatenated = v || 0x01 || priv_key || h1
+        concatenated = v.as_ref().to_vec();
+        concatenated.extend(one.as_ref().to_vec().into_iter());
+        let priv_key = unsafe {
+            slice::from_raw_parts(
+                pk.0.into_repr()
+                .as_ref()
+                .to_vec().as_ptr() as *const u8, 32
+            )
+        };
+        concatenated.extend(priv_key.to_vec().into_iter());
```

Comment:
For efficiency reasons use extend_from_slice to avoid large number of allocations